core/tests/test_utils.py

```diff
@@ -1838,6 +1843,123 @@ def swap(self, obj, attr, newvalue):
         finally:
             setattr(obj, attr, original)
 
+    @contextlib.contextmanager
+    def swap_with_checks(
+            self, obj, attr, newvalue, expected_args=None, expected_kwargs=None,
+            called=True, called_times=None,
+            expected_exception=IntendedException):
+        """Swap an object's function value within the context of a
+        'with' statement. The object can be anything that supports
+        getattr and setattr, such as class instances, modules, ...
+        Yields:
+            context: The context with function replaced.
+
+        Examples:
+            If you want to check subprocess.Popen is invoked twice
+            like `subprocess.Popen(['python'], shell=True)` and
+            `subprocess.Popen(['python2], shell=False), you can first
+            define the mock  function, then the swap, and just run the
+            target function in context, as follows:
+                def mock_popen(command, shell):
+                    return
+
+                popen_swap = self.swap_with_checks(
+                    subprocess, 'Popen', mock_popen, expected_args=[
+                        (['python'],), (['python2'],)], expected_kwargs={
+                            'shell': True,
+                            'shell': False
+                        })
+                with popen_swap:
+                    function_will_invoke_popen()
+
+        Args:
+            obj: Any. The target that you want to swap a function with.
+            attr: str. The name of the function to be swapped.
+            newvalue: Function. The new function you want to use.
+            expected_args: None|tuple|list[tuple]. The expected args that you
+                want this function to be invoked with. When its value is None,
+                args will  not be checked. If the value type is tuple, means
```

Comment:
Remove extra space after "will". (There are double spaces in odd places elsewhere, please remove.)

Reply:
Done